### PR TITLE
Fix bug in field level metadata matching code

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/8285

## Rationale for this change

Bug fix

## What changes are included in this PR?

What was happening is that the input to a join has metadata on one side, but not the other. When the inputs were reordered then metadata was not properly reordered as well

Specifically:
```
OutputRequirementExec
  ProjectionExec: expr=[id@0 as id]
    HashJoinExec: mode=Partitioned, join_type=Inner, on=[(id@0, id@0)]
      AggregateExec: mode=FinalPartitioned, gby=[id@0 as id], aggr=[]  <--******** this does not have metadata
        AggregateExec: mode=Partial, gby=[id@0 as id], aggr=[]
          UnionExec
            MemoryExec: partitions=1, partition_sizes=[1]
            MemoryExec: partitions=1, partition_sizes=[1]
      MemoryExec: partitions=1, partition_sizes=[1]                    <--******** this has metadata
```

So the output schema looks like:
```
  id (metadata)    <-- first input
  id (no metadata) <-- second input
```

When the join reorders the inputs, the result had the metadata incorrectly set:

```
  id (no metadata) <-- first input *** Metadata is switched ***
  id (metadata)    <-- second input
```

There is code in the join reordering that tries to remap the output so the schema is the same, however
it had a bug (was looking up field by name, not column index), so it picked the wrong column.

### CODE REVIEW

File: datafusion/sqllogictest/src/test_context.rs [hunk 1]
```diff
@@ -299,3 +299,31 @@ fn table_with_many_types() -> Arc<dyn TableProvider> {
     let provider = MemTable::try_new(Arc::new(schema), vec![vec![batch]]).unwrap();
     Arc::new(provider)
 }
+
+/// Registers a table_with_metadata that contains both field level and Table level metadata
+pub async fn register_metadata_tables(ctx: &SessionContext) {
```
Comment (alamb) [inline datafusion/sqllogictest/src/test_context.rs:304]: I intend to port several existing metadata tests to sqllogictest as well

File: datafusion/physical-plan/src/projection.rs [hunk 2]
```diff
@@ -257,16 +257,13 @@ fn get_field_metadata(
...
-    input_schema
-        .field_with_name(name)
-        .ok()
-        .map(|f| f.metadata().clone())
+    // Look up field by index in schema (not NAME as there can be more than one
+    // column with the same name)
+    e.as_any()
+        .downcast_ref::<Column>()
+        .map(|column| column.index())
+        .map(|idx| input_schema.field(idx).metadata())
```
Comment (comphead) [inline datafusion/physical-plan/src/projection.rs:265]: this can likely be done within a single `map`?

Comment (alamb) [reply datafusion/physical-plan/src/projection.rs]: good idea -- done in f5c557e1f8

**Comment by alamb** [comment]
Thank you for the review @comphead

**Pull request merged**